Support Write-Thru of EH variables in LSRA

CarolEidt opened the pull request:
Mark EH variables (those that are live in or out of exception regions) only as lvLiveInOutOfHndlr, not necessarily lvDoNotEnregister.
During register allocation, mark these as write-thru, and mark all defs as write-thru, ensuring that the stack value is always valid.
Mark those defs with GTF_SPILLED (this the "reload" flag and is not currently used for pure defs) to indicate that it should be kept in the register.
Mark blocks that enter EH regions as having no predecessor, and set the location of all live-in vars to be on the stack.
Change genFnPrologCalleeRegArgs to store EH vars also to the stack if they have a register assignment.

CarolEidt committed:
Support Write-Thru of EH variables in LSRA

Mark EH variables (those that are live in or out of exception regions) only as lvLiveInOutOfHndlr, not necessarily lvDoNotEnregister
During register allocation, mark these as write-thru, and mark all defs as write-thru, ensuring that the stack value is always valid.
Mark those defs with GTF_SPILLED (this the "reload" flag and is not currently used for pure defs) to indicate that it should be kept in the register.
Mark blocks that enter EH regions as having no predecessor, and set the location of all live-in vars to be on the stack.
Change genFnPrologCalleeRegArgs to store EH vars also to the stack if they have a register assignment.

CarolEidt committed:
Throughput tuning - rearrange some fields and short-circuit the physical register RefPositions during allocation.

CarolEidt commented:
@dotnet/jit-contrib PTAL

sandreenko commented on src/coreclr/src/jit/lsra.cpp:
```diff
@@ -4435,7 +4498,24 @@ void LinearScan::spillGCRefs(RefPosition* killRefPosition)
             continue;
         }
-        unassignPhysReg(regRecord, assignedInterval->recentRefPosition);
+        bool needsKill = varTypeIsGC(assignedInterval->registerType);
```
I think that was part was already merged in #679, why does github show that as a diff?

CarolEidt replied:
I believe it's because this is actually showing the original change that I extracted for that PR. I haven't yet pushed a rebased version of this PR.

sandreenko commented on src/coreclr/src/jit/lsrabuild.cpp:
```diff
@@ -2092,49 +2096,75 @@ void LinearScan::buildIntervals()
-                // Compute set difference: newLiveIn = currentLiveVars - predBlock->bbLiveOut
-                VarSetOps::DiffD(compiler, newLiveIn, predBlock->bbLiveOut);
-            }
-            bool needsDummyDefs = (!VarSetOps::IsEmpty(compiler, newLiveIn) && block != compiler->fgFirstBB);
```
Could you please explain how this PR relates to the deletion of `block != compiler->fgFirstBB` condition here?

CarolEidt replied:
With this PR, we have blocks with an incoming EH boundary that may have live-in candidate vars. Previously those vars would have no live-in candidate vars, by construction, so we didn't need to worry about whether they needed dummy defs. Now that we have live-in EH vars, we don't want dummy defs for those (they should always be on stack on entry), so they are treated as having no pred block (like `fgFirstBB`), and that becomes the right condition for determining when we don't want dummy defs.

sandreenko reviewed (commented):
I expect that change has an impact on throughput and memory consumption, how big is it?

The implementation looks good, but I am scared to have so many places where we check `writeThru`, that means it will be very easy to forget some of them.

CarolEidt commented:
> I expect that change has an impact on throughput and memory consumption, how big is it?

The Throughput tuning commit brings this from roughly a .05% throughput loss on x64 (in the noise on x86) to a .05% improvement on x64 and a .007% improvement (barely above the noise) on x86 for crossgen of SPC.dll

> The implementation looks good, but I am scared to have so many places where we check writeThru, that means it will be very easy to forget some of them.

I'm not sure I fully understand. This is pretty fundamental, as these variables need to be handled differently.

CarolEidt commented on src/coreclr/src/jit/lsra.h:
```diff
@@ -1672,11 +1682,17 @@ class Interval : public Referenceable
-    bool         isLocalVar : 1;
+
+    // Is this Interval currently in a register and live?
+    bool isActive;
```
I moved this field from the base type (`Referenceable`) to `Interval`, and changing it to a bitfield seemed to have a minor negative impact.

CarolEidt commented on src/coreclr/src/jit/compiler.h:
```diff
@@ -3015,7 +3014,11 @@ class Compiler
+    void lvaSetVarLiveInOutOfHandler(unsigned varNum);
     bool lvaVarDoNotEnregister(unsigned varNum);
+
+    bool lvaEnregEHVars;
```
We don't currently have any bitfields on `Compiler`, and given that there is only one instance per compilation, it seems that the efficiency of querying a byte overrides the storage impact.

CarolEidt commented on src/coreclr/src/jit/instr.cpp:
```diff
@@ -660,9 +660,18 @@ void CodeGen::inst_TT_RV(instruction ins, GenTree* tree, regNumber reg, unsigned
-
-    if (tree->gtFlags & GTF_SPILLED)
+    // Is this a spilled value?
+    bool isSpilledValue = ((tree->gtFlags & GTF_SPILLED) != 0);
```
Makes sense. With the removal of legacy jit, I'm pretty sure this code is only used for lclVar spilling, so I think I'll try to simplify it.

CarolEidt commented on src/coreclr/src/jit/lclvars.cpp:
```diff
@@ -85,6 +85,8 @@ void Compiler::lvaInit()
 
     structPromotionHelper = new (this, CMK_Generic) StructPromotionHelper(this);
+
+    lvaEnregEHVars = ((opts.compFlags & CLFLG_REGVAR) != 0);
```
Although I hadn't added it yet, I anticipated having a `COMPlus` variable to separately control this (which I'm hopefully about to push as an update to this PR). However, LSRA uses the condition above to set `enregisterLocalVars` so it wouldn't be possible to have `lvaEnregEHVars` be true and `enregisterLocalVars` be false.

CarolEidt commented on src/coreclr/src/jit/liveness.cpp:
```diff
@@ -2477,17 +2477,16 @@ void Compiler::fgInterBlockLocalVarLiveness()
+        // Mark all pointer variables live on exit from a 'finally' block as lvLiveInOutOfHndlr.
+        // and as 'explicitly initialized' (must-init) for GC-ref types.
 
         if (VarSetOps::IsMember(this, finallyVars, varDsc->lvVarIndex))
```
I don't understand how this could be changed to `else`.
Also, note that the structure has not fundamentally changed - I just changed the comments to use `//` since I was modifying the code nearby.

sandreenko replied:
The previous condition is:
`if (smth1) lvaSetVarLiveInOutOfHandler(varNum);` 
and this is: 
`if (smth2) lvaSetVarLiveInOutOfHandler(varNum);`
, so we can do 
`if (smth1 or smth2) lvaSetVarLiveInOutOfHandler(varNum);` 
or
```
if (smth1) lvaSetVarLiveInOutOfHandler(varNum);
else if (smth2) lvaSetVarLiveInOutOfHandler(varNum);
```

As I see from `lvaSetVarLiveInOutOfHandler` we don't need to call it twice, do we?

CarolEidt replied:
I see your point now. I'l restructure the conditions.

CarolEidt commented on src/coreclr/src/jit/codegencommon.cpp:
```diff
@@ -3815,7 +3825,7 @@ void CodeGen::genFnPrologCalleeRegArgs(regNumber xtraReg, bool* pXtraRegClobbere
 #endif // !_TARGET_64BIT_
         {
             // If not a stack arg go to the next one
```
How about `// If this arg is never on the stack, go to the next one.`?

CarolEidt commented on src/coreclr/src/jit/lsra.cpp:
```diff
@@ -847,11 +857,8 @@ void LinearScan::setBlockSequence()
-            if (((predBlock->bbFlags & BBF_KEEP_BBJ_ALWAYS) != 0) || (hasUniquePred && predBlock->hasEHBoundaryOut()))
+            if (predBlock->hasEHBoundaryOut())
             {
-                // Treat this as having incoming EH flow, since we can't insert resolution moves into
```
What this does is change the block to have no predecessor, so that it doesn't have to deal with mismatched locations of EH vars across the boundary. However, this change actually causes some small diffs when the EH write-thru is disabled, because it creates mismatches on the non-EH edges. I'm looking into how complex it would be to eliminate this change, but it may be excessively complex. The crossgen diffs over all jits & altjits for frameworks & tests shows only 4 bytes of diff each over two large methods for arm32, and 40 bytes of diff for one 4020 byte method for x64/ux altjit.

AndyAyersMS commented on src/coreclr/src/jit/lclvars.cpp:
```diff
@@ -4080,8 +4125,20 @@ void Compiler::lvaComputeRefCounts(bool isRecompute, bool setSlotNumbers)
+                        if (varDsc->lvLiveInOutOfHndlr && !varDsc->lvDoNotEnregister &&
+                            ((node->gtFlags & GTF_VAR_DEF) != 0))
+                        {
+                            varDsc->incRefCnts(0, this);
```
Is there some other way to accomplish this? I never like to see us bias this sort of accounting if we can avoid it.

Say down the road we want to use weighted ref counts for other optimizations -- for instance trying to allocate the most frequently accessed locals at small FP or SP offsets -- we'd want these weighted ref counts to reflect our best model of reality.

CarolEidt replied:
This is tricky because of the way we overload the ref counts for use by optimization and register allocation - the register allocator uses the weight to determine the value of allocating a register.
Note that this is only the `IsLir()` path.

AndyAyersMS commented on src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/AsyncMethodBuilderCore.cs:
```diff
@@ -25,39 +25,32 @@ internal static class AsyncMethodBuilderCore // debugger depends on this exact n
             // so that they won't "leak" out of the first await.
-            ExecutionContext? previousExecutionCtx = previousExecutionCtx0;
-            SynchronizationContext? previousSyncCtx = currentThread0._synchronizationContext;
+            ExecutionContext? previousExecutionCtx = currentThread._executionContext;
```
Would be good to post diffs for these (maybe in a gist). Also when these "manual write-thru" changes were added there was some kind of perf testing -- have you looked at trying to revisit those tests?

AndyAyersMS replied:
Ditto for the other Fx edits below...

CarolEidt replied:
I am unaware of  perf testing that was done when those mitigations were added - @stephentoub do you know how those might have been tested?

AndyAyersMS replied:
There are some perf tests in dotnet/coreclr#/15629.

AndyAyersMS reviewed (commented):
Some initial comments. Still need to look through lsra.cpp ...

AndyAyersMS commented:
Possible follow-up: wonder if we could now put live seems like this change might allow us to get rid of the `lvVolatileHint` computations we do in `lvaMarkLclRefs` and the code to copy volatile hinted locals in `optAddCopies`.

Also if you're interested in stress-testing, you might try the `mutate-test` tool from `jitutils`, it can introduce EH into all methods in a test case.

CarolEidt commented:
@AndyAyersMS - thanks. Right now I'm working on getting this checked in disabled. Turns out there are some minor regressions due to the way block boundaries are handled. I'm trying to get those eliminated. At that point it would be interesting to work on the stress testing; I think the other is best to leave for later.

AndyAyersMS commented on src/coreclr/src/jit/lsrabuild.cpp:
```diff
@@ -1181,18 +1181,22 @@ bool LinearScan::buildKillPositionsForNode(GenTree* tree, LsraLocation currentLo
-                regMaskTP newPreferences = allRegs(interval->registerType) & (~killMask);
 
-                if (newPreferences != RBM_NONE)
+                if (!interval->isWriteThru || !isCallKill)
```
Can you (perhaps in a comment) explain this part of the change?

AndyAyersMS commented on src/coreclr/src/jit/lsrabuild.cpp:
```diff
@@ -2317,6 +2359,62 @@ void LinearScan::buildIntervals()
+                    // so we won't break even until we have at least 4 * BB_UNITY_WEIGHT.
+                    // Given that we also don't have a good way to tell whether the variable is live
+                    // across a call in the non-EH code, we'll be extra conservative about this.
+                    // Note that for writeThru intervals we don't update the preferences to be only callee-save.
```
I take it this is the matching bit of logic for line 1185?

CarolEidt replied:
Yes, I'll add a pointer

AndyAyersMS commented:
Have now looked over everything but lsra.cpp; not sure how much help I can be with the changes there.

CarolEidt commented on src/coreclr/src/jit/codegencommon.cpp:
```diff
@@ -504,7 +504,10 @@ void CodeGenInterface::genUpdateRegLife(const LclVarDsc* varDsc, bool isBorn, bo
+        // If this is going live, the register must not have a variable in it, except
+        // in the case of an exception variable, which may be already treated as live
+        // in the register.
+        assert(varDsc->lvLiveInOutOfHndlr || ((regSet.GetMaskVars() & regMask) == 0));
```
I've pretty much left the assert\noway_assert distinction as-is - these are generally just either expanding or contracting the condition for an assert.

CarolEidt commented on src/coreclr/src/jit/compiler.h:
```diff
@@ -415,6 +415,8 @@ class LclVarDsc
     unsigned char lvFieldAccessed : 1;   // The var is a struct local, and a field of the variable is accessed.  Affects
                                          // struct promotion.
+    unsigned char lvLiveInOutOfHndlr : 1; // The variable is live in or out of an exception handler, and therefore must
+                                          // be on the stack (at least at those boundaries.)
```
I've added an " EH" annotation unconditionally, and left the 'H' character in the DNER dump.

CarolEidt committed:
Disable by default

CarolEidt commented:
@dotnet/jit-contrib - This is ready for review. It is disabled by default and there are zero diffs.

CarolEidt commented:
I plan to rebase and run jitstress jitstressregs once my two stress fixes are merged.

CarolEidt committed:
Clean up target macros, and fix lvMustInit of finally vars

CarolEidt commented:
/azp run runtime-coreclr jitstress2-jitstressregs

echesakovMSFT commented:
/azp list

lpereira commented on src/coreclr/src/jit/treelifeupdater.cpp:
```diff
@@ -139,31 +143,31 @@ void TreeLifeUpdater<ForCodeGen>::UpdateLifeVar(GenTree* tree)
                                 compiler->codeGen->genUpdateRegLife(fldVarDsc, isBorn, isDying DEBUGARG(tree));
                             }
-                            else
+                            if (isInMemory)
```
Would these changes from `else` to `if (isInMemory)` be better as `else if (isInMemory)` with another `else` with a `noway_assert()`?

CarolEidt replied:
No; the whole point of this change is that a variable can now be valid both in register and in memory, because when we "write-thru" a definition of an EH exposed variable it can remain live in the register as well.

CarolEidt commented:
@dotnet/jit-contrib ping

AndyAyersMS commented on src/coreclr/src/jit/codegenlinear.cpp:
```diff
@@ -1852,13 +1877,23 @@ void CodeGen::genProduceReg(GenTree* tree)
+
+            // The GTF_SPILL flag generally means that we need to spill this local.
+            // The exception is the case of an EH var that is being "spilled"
+            // to the stack, indicated by aby an EH GT_LCL_VAR use that is marked GTF_SPILL
```
Typo 'by aby'

AndyAyersMS commented on src/coreclr/src/inc/clrconfigvalues.h:
```diff
@@ -401,6 +401,7 @@ RETAIL_CONFIG_DWORD_INFO_EX(EXTERNAL_FeatureSIMD, W("FeatureSIMD"), EXTERNAL_Fea
 RETAIL_CONFIG_DWORD_INFO(INTERNAL_SIMD16ByteOnly, W("SIMD16ByteOnly"), 0, "Limit maximum SIMD vector length to 16 bytes (used by x64_arm64_altjit)")
 RETAIL_CONFIG_DWORD_INFO_EX(EXTERNAL_EnableAVX, W("EnableAVX"), EXTERNAL_JitEnableAVX_Default, "Enable AVX instruction set for wide operations as default", CLRConfig::REGUTIL_default)
 RETAIL_CONFIG_DWORD_INFO_EX(UNSUPPORTED_TrackDynamicMethodDebugInfo, W("TrackDynamicMethodDebugInfo"), 0, "Specifies whether debug info should be generated and tracked for dynamic methods", CLRConfig::REGUTIL_default)
+RETAIL_CONFIG_DWORD_INFO_EX(EXTERNAL_EnableEHWriteThru,  W("EnableEHWriteThru"), 0, "Enable enregistration of variables live on EH edges", CLRConfig::REGUTIL_default)
```
I don't think you need this, the runtime doesn't care...

CarolEidt replied:
Ah, yes - thanks.

AndyAyersMS commented on src/coreclr/src/jit/lsra.h:
```diff
@@ -78,6 +78,14 @@ inline regMaskTP calleeSaveRegs(RegisterType rt)
 }
 
+//------------------------------------------------------------------------
+// registerTypesEquivalent: Get the set of callee-save registers of the given RegisterType
```
Cut and paste issue in comment?

AndyAyersMS replied:
Looks like same just above too, in `calleeSaveRegs`

AndyAyersMS reviewed (commented):
Left a few more notes; I think there are still some unaddressed comments from earlier too.

CarolEidt committed:
PR Feedback

CarolEidt commented:
> I think there are still some unaddressed comments from earlier too.

I believe I addressed all of the previous comments, along with the latest; if not, could you point out what's left?

AndyAyersMS commented on src/coreclr/src/jit/codegenlinear.cpp:
```diff
@@ -1848,12 +1873,20 @@ void CodeGen::genProduceReg(GenTree* tree)
+            // by a non-def GT_LCL_VAR that is marked GTF_SPILL.
+            // This case occurs at the end of a block when its register is going dead.
+            // It is already valid on the stack.
+            if (((tree->gtFlags & GTF_VAR_DEF) != 0) || !varDsc->lvLiveInOutOfHndlr)
```
This still seems confusing, the comment is talking about `GTF_SPILL` but the code  you changed here does not look at this flag.

CarolEidt replied:
I've pushed another change to this comment, hopefully making it clear.

AndyAyersMS replied:
I think that's better, yes. @sandreenko do you agree?

AndyAyersMS commented:
> could you point out what's left?

Looks like just [this one](https://github.com/dotnet/runtime/pull/543#discussion_r357878689).

CarolEidt committed:
More PR feedback

CarolEidt committed:
Additional PR Feedback

CarolEidt commented:
I believe that I've addressed all the PR feedback.

AndyAyersMS reviewed (approved):
LGTM.

Bruce and I were talking about creating a rolling test where we could enable off by default changes like this so they don't regress while we're working on getting them to be on by default.
